### Creating, reading, and writing

In [1]:
import pandas as pd

**Creating data**

There are two core objects in pandas: the `DataFrame and the Series`.

A DataFrame is a table. It contains an array of individual entries, each of which has a certain value. Each entry corresponds with a row (or record) and a column.

For example, consider the following simple `DataFrame`:

In [2]:
pd.DataFrame({'Yes':['Love', 'Hate'], 'No':['Like', 'DisLike']})

No   Yes
0     Like  Love
1  DisLike  Hate

`DataFrame` entries are not limited to `string`. For instance, here's a DataFrame whose values are `inteeger`:

In [3]:
pd.DataFrame({'Yes':[90, 10], 'No':[79, 21]})

No  Yes
0  79   90
1  21   10

> We are using the pd.DataFrame constructor to generate these DataFrame objects. The syntax for declaring a new one is a dictionary whose keys are the column names, and whose values are a list of entries. This is the standard way of constructing a new DataFrame, and the one you are likliest to encounter.

> The dictionary-list constructor assigns values to the column labels, but just uses an ascending count from 0 (0, 1, 2, 3, ...) for the row labels. Sometimes this is OK, but oftentimes we will want to assign these labels ourselves.

> The list of row labels used in a DataFrame is known as an Index. We can assign values to it by using an index parameter in our constructor:

In [4]:
pd.DataFrame(
                {
                    'Alien':['I like her', 'She respect me'],
                    'Singhak':['She does not like me', 'She is too mature for me']
                },
                index = ['positive', 'negative'])

Alien                   Singhak
positive      I like her      She does not like me
negative  She respect me  She is too mature for me

In [5]:
pd.DataFrame(
{
    'Song:':['I love songs', 'I am not fond of songs'],
    'Talk':['Calm', 'Talkative'],
    'Height':[5.2, 5.7]
},
index = ['snigdha', 'anil'])

Height                   Song:       Talk
snigdha     5.2            I love songs       Calm
anil        5.7  I am not fond of songs  Talkative

> A Series, by contrast, is a sequence of data values. If a DataFrame is a table, a Series is a list. And in fact you can create one with nothing more than a list:

In [6]:
pd.Series([1,2,3,4,4,6])

0    1
1    2
2    3
3    4
4    4
5    6
dtype: int64

In [7]:
pd.Series(['Ram','Sita','Anil','Snigdha'])

0        Ram
1       Sita
2       Anil
3    Snigdha
dtype: object

In [8]:
pd.Series(['Ram','Sita','Anil','Snigdha'],
         index = ['I','n','f','o'],
        )

I        Ram
n       Sita
f       Anil
o    Snigdha
dtype: object

> A Series is, in essence, a single column of a DataFrame. So you can assign column values to the Series the same way as before, using an index parameter. However, a Series do not have a column name, it only has one overall name:

In [9]:
pd.Series([30, 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')

2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: int64

> Series and the DataFrame are intimately related. It's helpful to think of a DataFrame as actually being just a bunch of Series "glue together".

### Reading common file formats
> Being able to create a DataFrame and Series by hand is handy. But, most of the time, we won't actually be creating our own data by hand, we'll be working with data that already exists.

In [10]:
wine_reviews = pd.read_csv('winemag-data-130k-v2.csv')

find out the size/shape of csv data

In [11]:
wine_reviews.shape

(5525, 14)

Our data set has 5525 records and 14 colomns

We can examine the contents of the resultant DataFrame using the head command, which grabs the first five rows:

In [12]:
wine_reviews.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

> The pandas read_csv function is well-endowed, with over 30 optional parameters you can specify. For example, you can see in this dataset that the csv file has an in-built index, which pandas did not pick up on automatically. To make pandas use that column for the index (instead of creating a new one from scratch), we may specify and use an index_col

In [13]:
wine_reviews = pd.read_csv("winemag-data-130k-v2.csv", index_col=0)
wine_reviews.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

For now, let's move on to another common data format: SQL files.

SQL databases are where most of the data on the web ultimately gets stored. They can be used to store data on things as simple as recipes to things as complicated as "almost everything on the Kaggle website".

Connecting to a SQL database requires a lot more thought than reading from an Excel file. For one, you need to create a connector, something that will handle siphoning data from the database.

pandas won't do this for you automatically because there are many, many different types of SQL databases out there, each with its own connector.

In [14]:
import sqlite3
conn = sqlite3.connect('FPA_FOD_20170508.sqlite')

In [15]:
fires = pd.read_sql_query(
    sql='''
        Select *
        FROM fires
        LIMIT 100
        ''', con=conn)

In [16]:
fires.head()

OBJECTID  FOD_ID      FPA_ID SOURCE_SYSTEM_TYPE SOURCE_SYSTEM  \
0         1       1  FS-1418826                FED   FS-FIRESTAT   
1         2       2  FS-1418827                FED   FS-FIRESTAT   
2         3       3  FS-1418835                FED   FS-FIRESTAT   
3         4       4  FS-1418845                FED   FS-FIRESTAT   
4         5       5  FS-1418847                FED   FS-FIRESTAT   

  NWCG_REPORTING_AGENCY NWCG_REPORTING_UNIT_ID  NWCG_REPORTING_UNIT_NAME  \
0                    FS                USCAPNF    Plumas National Forest   
1                    FS                USCAENF  Eldorado National Forest   
2                    FS                USCAENF  Eldorado National Forest   
3                    FS                USCAENF  Eldorado National Forest   
4                    FS                USCAENF  Eldorado National Forest   

  SOURCE_REPORTING_UNIT SOURCE_REPORTING_UNIT_NAME  \
0                  0511     Plumas National Forest   
1                  0503   Eldorado National Forest   
2                  0503   Eldorado National Forest   
3                  0503   Eldorado National Forest   
4                  0503   Eldorado National Forest   

                         ...                         FIRE_SIZE_CLASS  \
0                        ...                                       A   
1                        ...                                       A   
2                        ...                                       A   
3                        ...                                       A   
4                        ...                                       A   

    LATITUDE   LONGITUDE OWNER_CODE       OWNER_DESCR STATE COUNTY FIPS_CODE  \
0  40.036944 -121.005833        5.0              USFS    CA     63       063   
1  38.933056 -120.404444        5.0              USFS    CA     61       061   
2  38.984167 -120.735556       13.0  STATE OR PRIVATE    CA     17       017   
3  38.559167 -119.913333        5.0              USFS    CA      3       003   
4  38.559167 -119.933056        5.0              USFS    CA      3       003   

   FIPS_NAME                                              Shape  
0     Plumas  b'\x00\x01\xad\x10\x00\x00\xe8d\xc2\x92_@^\xc0...  
1     Placer  b'\x00\x01\xad\x10\x00\x00T\xb6\xeej\xe2\x19^\...  
2  El Dorado  b'\x00\x01\xad\x10\x00\x00\xd0\xa5\xa0W\x13/^\...  
3     Alpine  b'\x00\x01\xad\x10\x00\x00\x94\xac\xa3\rt\xfa]...  
4     Alpine  b'\x00\x01\xad\x10\x00\x00@\xe3\xaa.\xb7\xfb]\...  

[5 rows x 39 columns]

### Writing common file formats

Writing data to a file is usually easier than reading it out of one, because pandas handles the nuisance of conversions for you.

We'll start with CSV files again. The opposite of from_csv, which reads our data, is to_csv, which writes it. With CSV files it's dead simple:

In [17]:
wine_reviews.head(100).to_csv('wine_review.csv')

To write an Excel file back you need to_excel and the sheet_name again:

In [18]:
wine_reviews.head().to_excel('wine_review.xlsx', sheet_name="wine")

to output to a SQL database, supply the name of the table in the database we want to throw the data into, and a connector:

In [19]:
conn = sqlite3.connect("fires.sqlite")
fires.head(50).to_sql("fires", conn)

Dataset refrence

https://www.kaggle.com/residentmario/creating-reading-and-writing-reference/data